0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,77 @@
+# Pre-Shared Key
+
+The Pre-Shared Key (PSK) protocol is an end-to-end transport protocol, used by the sender and receiver of an ILP payment to decide on a condition and fulfillment for a payment. The protocol also encrypts any additional data sent along with the payment. The encryption algorithm used is AES-256-CTR.
```

Comment:
The HMAC of the whole packet (including the account, amount, unencrypted headers, and AES-CTR encrypted blob) is used as the condition fulfillment, ensuring that if the packet is modified in any way the recipient will not generate the correct fulfillment and the payment will be rejected. That said, your input on the key generation procedure would be very welcome.

**Receiver setup** (pre-PSK -- this is an implementation-specific flow)
```
receiver_secret = rand(32)
token = rand(16)
receiver_id = HMAC(receiver_secret, "ilp_psk_receiver_id").slice(0,8)
receiver_address = ilp_address + "." + receiver_id + base64urlencode(token)
shared_secret = HMAC(HMAC(receiver_secret, "ilp_psk_generation"), token)
```
Receiver gives the `receiver_address` and `shared_secret` to the sender over an authenticated, encrypted communication channel.

**Sender side**
```
nonce = rand(16)
encryption_key = HMAC(shared_secret, nonce) // TODO should this be generated with another constant string as well as the nonce?
private_request = private_headers + data
encrypted_data = AES-256-CTR(encryption_key, private_request)
public_request = "PSK 1.0..." + "Key: hmac-sha-256 " + nonce + public_headers + psk_encrypted_data // nonce is sent in cleartext
ilp_packet = Packet({ amount: amount, account: receiver_address, data: public_request})
fulfillment = HMAC(HMAC(shared_secret, "ilp_psk_condition"), ilp_packet)
condition = SHA-256(fulfillment)
```

**Receiver side**
```
token = base64decode(ilp_packet.account.slice(-11))
shared_secret = HMAC(HMAC(receiver_secret, "ilp_psk_generation"), token)
nonce = packet.data.key
decryption_key = HMAC(shared_secret, nonce)
fulfillment = HMAC(HMAC(shared_secret, "ilp_psk_condition"), ilp_packet)
```